keps/sig-apps/3939-include-terminating-pods-as-active/README.md

```diff
@@ -0,0 +1,811 @@
+for more examples.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Open Questions on Deployment Controller
+
+The Deployment API is open for discussion.  We put the field in Deployment/ReplicaSet because it is related to RolloutStrategy.
+It is not clear if `recreate` and/or `rollingupdate` need this API for both rollout options.
+
+Another open question is if we want to include Deployments in the initial release of this feature.  There is some discussion about releasing the Job API first and then follow up with Deployment.  
+
+We decided to define the APIs in this KEP as they can utilize the same implementation.
+
+#### Open Questions on Job Controller
+
+With 3329-retriable-and-non-retriable-failures and PodFailurePolicy enabled, terminating pods are only marked as failed once they have been transitioned to failed.  If PodFailurePolicy is disabled, then we mark a terminating pod as failed as soon as deletion is registered.  
+
+Should we add a new field to the status that reflects terminating pods?
+
+[Job controller should wait for Pods to be in a terminal phase before considering them failed or succeeded](https://github.com/kubernetes/kubernetes/issues/116858) is a relevant issue for this case.  
+I am not sure how to handle these two different cases if we want to count terminating pods as active.  
+
+Should we use this feature to help solve 116858?  When this feature toggle is on, then we mark terminating pods only as failed once they are complete regardless of PodFailurePolicy.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Name Choices
+
+- TerminatingAsActive
+- ActiveUntilTerminal
+- DelayPodRecreationUntilTerminal
+- ?
+
+### Job API Definition
+
+At the JobSpec level, we are adding a new BoolPtr field:
+
+```golang
+type JobSpec struct{
+  ...
+ // terminatingAsActive specifies if the Job controller should include terminating pods
+ // as active. If the field is true, then the Job controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
+}
+```
+
+### Deployment/ReplicaSet API
+
+```golang
+// DeploymentStrategy stores information about the strategy and rolling-update
+// behavior of a deployment.
+type DeploymentStrategy struct {
+  ... 
+  // TerminatingAsActive specifies if the Deployments should include terminating pods
+ // as active. If the field is true, then the Deployment controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
```

Comment:
I like this because replacement is the action (but if we add scale we might need a new name) - do we have other variations beyond these two?   Would we ever have compositions of behavior?

Reply:
I think it could be argued that scale also replaces pods, albeit only the terminating ones.

I am not aware of any other variations and I am not sure that composition with e.g. scaling (`TerminalScalingTerminatingDeployment` or reverse) is useful to anyone. 

Nevertheless, I could imagine a new feature where a pod could signal that it wants to be replaced via conditions, when it releases an important resource. But it still needs some time for a cleanup. This is just to illustrate that we cannot be sure that boolean will be sufficient in the future.

Reply:
It would be also good to consider the StatefulSet and its field `PodManagementPolicy` as a name here, because it is very similar to this. The problem is that it also controls how many pods will get updated, so the semantics differ a bit. And we are not trying to manage the pod in full context, just how it gets deleted. Not sure if it makes sense to go for a more generic name?

Reply:
+1, except that in Job we cannot say `Terminal`, because a Pod marked as Succeeded is not recreated, only Failed.

I would consider splitting the changes to Deployment in a separate KEP.

Reply:
split into two different keps.